src/back/spv/writer.rs

```diff
@@ -1320,6 +1355,234 @@ impl Writer {
+            // during validation, so there's no need to check anything here.
+            if let crate::Expression::Constant(_) = ir_function.expressions[index] {
+                return Ok(index_id);
+            }
+
+            let limit = known_length - 1;
+            self.get_index_constant(limit)?
+        } else {
+            self.write_runtime_array_limit(array, ir_module, ir_function, function, block)?
+        };
+
+        let restricted_index_id = self.id_gen.next();
+        block.body.push(Instruction::ext_inst(
+            self.gl450_ext_inst_id,
+            spirv::GLOp::UMin,
+            index_type_id,
+            restricted_index_id,
+            &[index_id, limit_id],
+        ));
+        Ok(restricted_index_id)
+    }
+
+    /// Write an array index bounds comparison to `block`.
+    ///
+    /// If we performed the bounds check at validation time, generate no code
+    /// and return `Ok(None)`, to indicate that no branch is necessary.
+    ///
+    /// Otherwise, return `Ok(Some(id))`, where `id` is a boolean value that is
+    /// true if `index` is in bounds for `array`.
+    #[allow(clippy::too_many_arguments)]
+    fn write_index_comparison(
+        &mut self,
+        array: Handle<crate::Expression>,
+        index: Handle<crate::Expression>,
+        ir_module: &crate::Module,
+        ir_function: &crate::Function,
+        fun_info: &FunctionInfo,
+        function: &mut Function,
+        block: &mut Block,
+    ) -> Result<Option<Word>, Error> {
+        let array_ty = fun_info[array].ty.inner_with(&ir_module.types);
+        let index_id = self.cached[index];
+
+        // Get the array's length. Return early if we've already done the bounds check.
+        let length_id = if let Some(known_length) = array_ty.to_known_length(ir_module) {
+            // We should have thrown out all attempts to subscript zero-length
+            // sequences during validation.
+            assert!(known_length > 0);
+
+            // When the length and index are known at compile time, we do the bounds check
+            // during validation, so there's no need to check anything here.
+            if let crate::Expression::Constant(_) = ir_function.expressions[index] {
+                return Ok(None);
+            }
+
+            self.get_index_constant(known_length)?
+        } else {
+            self.write_runtime_array_length(array, ir_module, ir_function, function, block)?
+        };
+
+        // Construct the `bool` type.
+        let bool_type = LocalType::Value {
+            vector_size: None,
+            kind: crate::ScalarKind::Bool,
+            width: 1,
+            pointer_class: None,
+        };
+        let bool_type_id = self.get_type_id(bool_type.into())?;
+
+        // Compare the index against the length.
+        let condition_id = self.id_gen.next();
+        block.body.push(Instruction::binary(
+            spirv::Op::ULessThan,
```

Comment:
A bit of a problem here: the `write_runtime_array_length` result is actually "length(xx) - 1". So comparing with `<` is incorrect with it.

Reply:
I don't think that's correct - `write_runtime_array_length` is the length of the array; it's the `_limit` functions that return one less than the length. I'll change the name of the `_limit` functions to `_max_index`, to make this clearer.